graphql_api/helpers/connection.py

```diff
@@ -68,6 +69,110 @@ def page_info(self, *args, **kwargs):
         }
 
 
+class ArrayPaginator:
+    """Cursor-based paginator for in-memory arrays."""
+
+    def __init__(
+        self,
+        data: List[Any],
+        first: Optional[int] = None,
+        last: Optional[int] = None,
+        after: Optional[str] = None,
+        before: Optional[str] = None,
+    ):
+        self.data = data
+        self.start_index = 0
+        self.end_index = len(data)
+
+        if first and last:
+            raise ValueError("Cannot provide both 'first' and 'last'")
+
+        if after is not None:
+            self.start_index = int(after) + 1
+
+        if before is not None:
+            self.end_index = min(self.end_index, int(before))
+
+        # Ensure valid bounds after 'after' and 'before'
+        self.start_index = max(self.start_index, 0)
+        self.end_index = min(self.end_index, len(data))
+
+        if first is not None:
+            self.end_index = min(self.start_index + first, len(data))
+
+        if last is not None:
+            range_length = self.end_index - self.start_index
+            if range_length > last:
+                self.start_index = self.end_index - last
+
+        # Ensure bounds remain valid
+        self.start_index = max(self.start_index, 0)
+        self.end_index = min(self.end_index, len(data))
+
+    def cursor(self, position: int) -> str:
+        """Generate a cursor based on the position (index)."""
+        return str(position)
+
+    @property
+    def page(self) -> List[Any]:
+        """Returns the sliced page of data."""
+        return self.data[self.start_index : self.end_index]
+
+    @property
+    def has_next(self) -> bool:
+        """Check if there's a next page."""
+        return self.end_index < len(self.data)
+
+    @property
+    def has_previous(self) -> bool:
+        """Check if there's a previous page."""
+        return self.start_index > 0
+
+
+class ArrayConnection:
+    """Connection wrapper for array pagination."""
+
+    def __init__(self, data: List[Any], paginator: ArrayPaginator, page: List[Any]):
```

Comment:
You don't actually need `data` and `page` right?
Total count will be `len(self.paginator.data)` and references of `self.page` will be `self.paginator.page`.
IMO adding these two params to the `ArrayConnection` class just adds more confusion

Reply:
You're right, I was trying to decouple the two to some extent, but it's not necessary at this point, as I doubt we'll use it anywhere besides the `query_to_connection` function